client/settings/advanced-settings/index.js

```diff
@@ -1,76 +1,34 @@
 /**
  * External dependencies
  */
-import { useState, useCallback, useEffect, useRef } from '@wordpress/element';
+import React, { useContext } from 'react';
+import { useEffect, useRef } from '@wordpress/element';
 import { Icon, chevronDown, chevronUp } from '@wordpress/icons';
-import { __, sprintf } from '@wordpress/i18n';
-import {
-	Card,
-	CardBody,
-	CheckboxControl,
-	TextControl,
-	TextareaControl,
-	Notice,
-	Button,
-	ExternalLink,
-} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { Card, CardBody, Button } from '@wordpress/components';
 
 /**
  * Internal dependencies
  */
 import SettingsSection from '../settings-section';
+import BlockAppearance from './block-appearance';
+import DebugMode from './debug-mode';
+import useToggle from './use-toggle';
 import './style.scss';
-import { useDebugLog, useDevMode } from '../../data';
-
-const useToggle = ( initialValue = false ) => {
-	const [ value, setValue ] = useState( initialValue );
-	const toggleValue = useCallback(
-		() => setValue( ( oldValue ) => ! oldValue ),
-		[ setValue ]
-	);
-
-	return [ value, toggleValue ];
-};
-
-const CheckboxToggle = ( { label, defaultIsChecked, children } ) => {
-	const [ isExpanded, toggleIsExpanded ] = useToggle( defaultIsChecked );
-	const wrapperRef = useRef( null );
-
-	useEffect( () => {
-		if ( ! isExpanded ) return;
-		if ( ! wrapperRef.current ) return;
-
-		const input = wrapperRef.current.querySelector( 'input, textarea' );
-		if ( ! input ) return;
-
-		input.focus();
-	}, [ isExpanded ] );
-
-	return (
-		<>
-			<CheckboxControl
-				label={ label }
-				checked={ isExpanded }
-				onChange={ toggleIsExpanded }
-			/>
-			{ isExpanded && <div ref={ wrapperRef }>{ children }</div> }
-		</>
-	);
-};
+import WCPaySettingsContext from '../wcpay-settings-context';
 
 const AdvancedSettings = () => {
-	const isDevModeEnabled = useDevMode();
 	const [ isSectionExpanded, toggleIsSectionExpanded ] = useToggle( false );
-	const firstHeadingElementRef = useRef( null );
-	const [ isLoggingChecked, setIsLoggingChecked ] = useDebugLog();
-	const [ customFontValue, setCustomFontValue ] = useState( '' );
-	const [ cssStylingValue, setCssStylingValue ] = useState( '' );
+	const advancedSectionContainerRef = useRef( null );
+	const {
+		featureFlags: { upe: isUpeEnabled },
+	} = useContext( WCPaySettingsContext );
 
 	useEffect( () => {
 		if ( ! isSectionExpanded ) return;
-		if ( ! firstHeadingElementRef.current ) return;
+		if ( ! advancedSectionContainerRef.current ) return;
 
-		firstHeadingElementRef.current.focus();
+		advancedSectionContainerRef.current.focus();
```

Comment:
The focus should really go to the heading element - otherwise it's not going to be correctly announced to screen readers.
Setting the focus on the wrapper `div` element won't do much, I'm afraid.

With this refactor, you can not just do a `useEffect` on mount in the `DebugMode` component to focus on its heading.


```
diff --git a/client/settings/advanced-settings/debug-mode.js b/client/settings/advanced-settings/debug-mode.js
index ce4c41ff..4d494849 100644
--- a/client/settings/advanced-settings/debug-mode.js
+++ b/client/settings/advanced-settings/debug-mode.js
@@ -3,6 +3,7 @@
  */
 import { CheckboxControl } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import { useEffect, useRef } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -12,10 +13,19 @@ import { useDebugLog, useDevMode } from 'data';
 const DebugMode = () => {
 	const isDevModeEnabled = useDevMode();
 	const [ isLoggingChecked, setIsLoggingChecked ] = useDebugLog();
+	const headingRef = useRef( null );
+
+	useEffect( () => {
+		if ( ! headingRef.current ) return;
+
+		headingRef.current.focus();
+	}, [] );
 
 	return (
 		<>
-			<h4>{ __( 'Debug mode', 'woocommerce-payments' ) }</h4>
+			<h4 ref={ headingRef } tabIndex="-1">
+				{ __( 'Debug mode', 'woocommerce-payments' ) }
+			</h4>
 			<CheckboxControl
 				label={
 					isDevModeEnabled
diff --git a/client/settings/advanced-settings/index.js b/client/settings/advanced-settings/index.js
index 4fa33dbc..af93fa5f 100644
--- a/client/settings/advanced-settings/index.js
+++ b/client/settings/advanced-settings/index.js
@@ -2,7 +2,6 @@
  * External dependencies
  */
 import React, { useContext } from 'react';
-import { useEffect, useRef } from '@wordpress/element';
 import { Icon, chevronDown, chevronUp } from '@wordpress/icons';
 import { __ } from '@wordpress/i18n';
 import { Card, CardBody, Button } from '@wordpress/components';
@@ -19,18 +18,10 @@ import WCPaySettingsContext from '../wcpay-settings-context';
 
 const AdvancedSettings = () => {
 	const [ isSectionExpanded, toggleIsSectionExpanded ] = useToggle( false );
-	const advancedSectionContainerRef = useRef( null );
 	const {
 		featureFlags: { upe: isUpeEnabled },
 	} = useContext( WCPaySettingsContext );
 
-	useEffect( () => {
-		if ( ! isSectionExpanded ) return;
-		if ( ! advancedSectionContainerRef.current ) return;
-
-		advancedSectionContainerRef.current.focus();
-	}, [ isSectionExpanded ] );
-
 	return (
 		<>
 			<SettingsSection>
@@ -45,13 +36,8 @@ const AdvancedSettings = () => {
 				<SettingsSection>
 					<Card>
 						<CardBody size="large">
-							<div
-								ref={ advancedSectionContainerRef }
-								tabIndex="-1"
-							>
-								<DebugMode />
-								{ isUpeEnabled && <BlockAppearance /> }
-							</div>
+							<DebugMode />
+							{ isUpeEnabled && <BlockAppearance /> }
 						</CardBody>
 					</Card>
 				</SettingsSection>
```

Reply:
Awesome!